File: 1ra Forma/GlobalResponse.java
// Framework: Spring Boot
// Archivo: GlobalResponse.java

package com.example.common.models;

import com.fasterxml.jackson.annotation.JsonInclude;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;

/**
 * Respuesta estándar para operaciones de API en Spring Boot
 */
@Data
@Builder
@NoArgsConstructor
@AllArgsConstructor
@JsonInclude(JsonInclude.Include.NON_NULL)
public class GlobalResponse<T> {
    
    private Boolean success;
    private String message;
    private String code;
    private T data;
    private Integer totalRows;
    private Integer currentPage;
    private Integer totalPages;
    private Integer pageSize;

    /**
     * Crea una respuesta exitosa simple
     */
    public static <T> GlobalResponse<T> ok(String message) {
        return GlobalResponse.<T>builder()
                .success(true)
                .message(message)
                .build();
    }

    /**
     * Crea una respuesta exitosa con datos
     */
    public static <T> GlobalResponse<T> okWithData(T data, String message, String code) {
        return GlobalResponse.<T>builder()
                .success(true)
                .data(data)
                .message(message)
                .code(code)
                .build();
    }

    /**
     * Crea una respuesta exitosa con datos (sin código)
     */
    public static <T> GlobalResponse<T> okWithData(T data, String message) {
        return okWithData(data, message, null);
    }

    /**
     * Crea una respuesta exitosa con paginación
     */
    public static <T> GlobalResponse<T> okPaginated(
            T data,
            int totalRows,
            int currentPage,
            int pageSize,
            String message,
            String code) {
        
        int totalPages = pageSize > 0 ? (int) Math.ceil((double) totalRows / pageSize) : 0;
        
        return GlobalResponse.<T>builder()
                .success(true)
                .data(data)
                .totalRows(totalRows)
                .currentPage(currentPage)
                .totalPages(totalPages)
                .pageSize(pageSize)
                .message(message)
                .code(code)
                .build();
    }

    /**
     * Crea una respuesta de error
     */
    public static <T> GlobalResponse<T> error(String message, String code) {
        return GlobalResponse.<T>builder()
                .success(false)
                .message(message)
                .code(code)
                .build();
    }

    /**
     * Crea una respuesta de error con datos
     */
    public static <T> GlobalResponse<T> errorWithData(T data, String message, String code) {
        return GlobalResponse.<T>builder()
                .success(false)
                .data(data)
                .message(message)
                .code(code)
                .build();
    }
}

// Ejemplo de uso en Spring Boot
/*
@RestController
@RequestMapping("/api/users")
public class UserController {

    @GetMapping
    public ResponseEntity<GlobalResponse<List<User>>> getUsers() {
        List<User> users = Arrays.asList(
            new User(1, "Juan"),
            new User(2, "María")
        );
        return ResponseEntity.ok(
            GlobalResponse.okWithData(users, "Usuarios obtenidos")
        );
    }

    @GetMapping("/paginated")
    public ResponseEntity<GlobalResponse<List<User>>> getUsersPaginated() {
        List<User> users = IntStream.range(1, 11)
            .mapToObj(i -> new User(i, "User" + i))
            .collect(Collectors.toList());
        
        return ResponseEntity.ok(
            GlobalResponse.okPaginated(users, 100, 1, 10, "Usuarios obtenidos", null)
        );
    }
}
*/

File: 1ra Forma/global_response.go
// Framework: Gin (Go)
// Archivo: global_response.go

package models

import "math"

// GlobalResponse representa la respuesta estándar de API
type GlobalResponse[T any] struct {
    Success     bool   `json:"success"`
    Message     string `json:"message,omitempty"`
    Code        string `json:"code,omitempty"`
    Data        *T     `json:"data,omitempty"`
    TotalRows   int    `json:"totalRows,omitempty"`
    CurrentPage int    `json:"currentPage,omitempty"`
    TotalPages  int    `json:"totalPages,omitempty"`
    PageSize    int    `json:"pageSize,omitempty"`
}

// Ok crea una respuesta exitosa simple
func Ok[T any](message string) *GlobalResponse[T] {
    return &GlobalResponse[T]{
        Success: true,
        Message: message,
    }
}

// OkWithData crea una respuesta exitosa con datos
func OkWithData[T any](data T, message, code string) *GlobalResponse[T] {
    return &GlobalResponse[T]{
        Success: true,
        Data:    &data,
        Message: message,
        Code:    code,
    }
}

// OkPaginated crea una respuesta exitosa con paginación
func OkPaginated[T any](data T, totalRows, currentPage, pageSize int, message, code string) *GlobalResponse[T] {
    totalPages := 0
    if pageSize > 0 {
        totalPages = int(math.Ceil(float64(totalRows) / float64(pageSize)))
    }

    return &GlobalResponse[T]{
        Success:     true,
        Data:        &data,
        TotalRows:   totalRows,
        CurrentPage: currentPage,
        TotalPages:  totalPages,
        PageSize:    pageSize,
        Message:     message,
        Code:        code,
    }
}

// Error crea una respuesta de error
func Error[T any](message, code string) *GlobalResponse[T] {
    return &GlobalResponse[T]{
        Success: false,
        Message: message,
        Code:    code,
    }
}

// ErrorWithData crea una respuesta de error con datos
func ErrorWithData[T any](data *T, message, code string) *GlobalResponse[T] {
    return &GlobalResponse[T]{
        Success: false,
        Data:    data,
        Message: message,
        Code:    code,
    }
}

// Ejemplo de uso en Gin
/*
package main

import (
    "github.com/gin-gonic/gin"
    "net/http"
)

type User struct {
    ID   int    `json:"id"`
    Name string `json:"name"`
}

func main() {
    r := gin.Default()

    r.GET("/api/users", func(c *gin.Context) {
        users := []User{
            {ID: 1, Name: "Juan"},
            {ID: 2, Name: "María"},
        }
        response := OkWithData(users, "Usuarios obtenidos", "")
        c.JSON(http.StatusOK, response)
    })

    r.GET("/api/users/paginated", func(c *gin.Context) {
        users := make([]User, 10)
        for i := 0; i < 10; i++ {
            users[i] = User{ID: i + 1, Name: fmt.Sprintf("User%d", i+1)}
        }
        response := OkPaginated(users, 100, 1, 10, "Usuarios obtenidos", "")
        c.JSON(http.StatusOK, response)
    })

    r.Run(":8080")
}
*/